baseServer.go

```diff
@@ -0,0 +1,216 @@
+	return serv.agentMap
+}
+
+func (serv *BaseServer[T]) agentStoppedTalking(id uuid.UUID) {
+	serv.agentMapRWMutex.Lock()
+	fmt.Println("sending stop talking request,id:", id)
+	serv.agentStoppedTalkingMap[id] = struct{}{}
+
+	if len(serv.agentStoppedTalkingMap) == len(serv.agentMap) {
+		serv.doneChannelOnce.Do(func() {
+			select {
+			case serv.messagingFinished <- struct{}{}:
+				close(serv.messagingFinished)
+			default:
+			}
+		})
+	}
+	serv.agentMapRWMutex.Unlock()
+}
+
+func (serv *BaseServer[T]) SetRunHandler(handler RoundRunner) {
+	serv.roundRunner = handler
+}
+
+func (serv *BaseServer[T]) checkHandler() {
+	if serv.roundRunner == nil {
+		panic("round running handler has not been set. Have you run SetRunHandler?")
+	}
+}
+
+func (serv *BaseServer[T]) RunTurn() {}
+
+func (bs *BaseServer[T]) RemoveAgent(agentToRemove T) {
+	delete(bs.agentMap, agentToRemove.GetID())
+}
+
+func (bs *BaseServer[T]) GetIterations() int {
+	return bs.iterations
+}
+
+func (bs *BaseServer[T]) RunRound() {}
+
+type AgentGenerator[T IAgent[T]] func(IExposedServerFunctions[T]) T
+
+type AgentGeneratorCountPair[T IAgent[T]] struct {
+	generator AgentGenerator[T]
+	count     int
+}
+
+func MakeAgentGeneratorCountPair[T IAgent[T]](generatorFunction AgentGenerator[T], count int) AgentGeneratorCountPair[T] {
+	return AgentGeneratorCountPair[T]{
+		generator: generatorFunction,
+		count:     count,
+	}
+}
+
+func (bs *BaseServer[T]) GenerateAgentArrayFromMap() []T {
+
+	agentMapToArray := make([]T, len(bs.agentMap))
+
+	i := 0
+	for _, ag := range bs.agentMap {
+		agentMapToArray[i] = ag
+		i++
+	}
+	return agentMapToArray
+}
+
+func (bs *BaseServer[T]) SendSynchronousMessage(msg IMessage[T], recipients []uuid.UUID) {
+	for _, recip := range recipients {
+		fmt.Println(recip, msg.GetSender())
+		if msg.GetSender() == recip {
+			continue
+		} else {
+			msg.InvokeMessageHandler(bs.agentMap[recip])
+		}
+	}
+
+}
+
+func (bs *BaseServer[T]) RunSynchronousMessagingSession() {
+	for _, agent := range bs.agentMap {
+		agent.RunSynchronousMessaging()
+	}
+}
+
+func (bs *BaseServer[T]) initialiseAgents(m []AgentGeneratorCountPair[T]) {
+
+	for _, pair := range m {
+		for i := 0; i < pair.count; i++ {
+			agent := pair.generator(bs)
+			bs.AddAgent(agent)
+		}
+	}
+
+}
+
+// generate a server instance based on a mapping function and number of iterations
+func CreateServer[T IAgent[T]](generatorArray []AgentGeneratorCountPair[T], iterations, turns int, turnMaxDuration time.Duration) *BaseServer[T] {
+	serv := &BaseServer[T]{
+		agentMap:               make(map[uuid.UUID]T),
+		agentIdSet:             make(map[uuid.UUID]struct{}),
+		listeningWaitGroup:     &sync.WaitGroup{},
+		agentStoppedTalkingMap: make(map[uuid.UUID]struct{}),
+		turnTimeout:            turnMaxDuration,
+		roundRunner:            nil,
+		iterations:             iterations,
+		turns:                  turns,
+		messagingFinished:      make(chan struct{}, 1),
```

Comment:
let's make this unbuffered to ensure that we don't miss the signal from either the agents or failsafe